proofs/erc20.md

```diff
@@ -0,0 +1,1030 @@
+      _:List
+    (
+      .List
+    =>
+      ListItem({ ACCT_ID
+               | #parseHexWord(Keccak256(#generateSignature("Approval", #address(CALLER_ID), #address(SPENDER), #uint256(VALUE))))
+               : CALLER_ID
+               : SPENDER
+               : .WordStack
+               | #asByteStackInWidth(VALUE, 32)
+               })
+    )
+```
+
+`<refund>` cell specifies that a refund may be issued. This function will refund a gas if the allowance value is 0 while the previous value is greater than 0, re-claiming the corresponding entry of the storage.
+
+Note that, however, we have not specified the refund detail since it is not essential for the functional correctness. We can specify that upon request.
+
+```{.k .erc20-spec .hkg-spec}
+refund: _ => _
+```
+
+`<storage>` cell specifies that the value of `allowances[CALLER_ID][SPENDER]` will be updated to `VALUE` after the execution.
+
+```{.k .erc20-spec .hkg-spec}
+storage:
+    #hashedLocation({lang}, {pos(allowances)}, CALLER_ID SPENDER) |-> (_:Int => VALUE)
+    _:Map
+```
+
+Unlike the ERC20-K specification, we do not specify the case when `VALUE` is less than 0 because it is not possible --- the `VALUE` parameter is of type `uint256`, an *unsigned* 256-bit integer.
+
+Indeed, the ABI call mechanism will reject a call to this function if the `VALUE` is negative, which is out of the scope of the EVM-level specification since it happens in the network level outside the EVM execution.
+
+```{.k .erc20-spec .hkg-spec}
+requires:
+    andBool 0 <=Int SPENDER   andBool SPENDER   <Int (2 ^Int 160)
+    andBool 0 <=Int VALUE     andBool VALUE     <Int (2 ^Int 256)
+
+comments:
+    // NOTE: negative VALUE is not possible since it is of `num256` type
+```
+
+
+
+
+transfer
+--------
+
+`transfer` function transfers the amount of VALUE from the caller to the account TO_ID.
+
+```{.k .erc20-spec .hkg-spec}
+[transfer]
+
+callData: #abiCallData("transfer", #address(TO_ID), #uint256(VALUE))
+
+gas: {gascap} => _
+```
+
+`transfer` function admits two types of behaviors: it either
+- succeeds in transferring the value and returns true (i.e., 1), or
+- fails due to the arithmetic overflow and throws an exception.
+
+We present two specifications, one for each case.
+
+### transfer success
+
+```{.k .erc20-spec .hkg-spec}
+[transfer-success]
+
+k: #execute => (RETURN RET_ADDR:Int 32 ~> _)
+
+localMem:
+      .Map
+    =>
+    (
+      .Map[ RET_ADDR := #asByteStackInWidth(1, 32) ]
+      _:Map
+    )
+
+```
+
+`<log>` cell specifies that it generates an event log of type `Transfer`.
+
+```{.k .erc20-spec .hkg-spec}
+log:
+      _:List
+    (
+      .List
+    =>
+      ListItem({ ACCT_ID
+               | #parseHexWord(Keccak256(#generateSignature("Transfer", #address(CALLER_ID), #address(TO_ID), #uint256(VALUE))))
+               : CALLER_ID
+               : TO_ID
+               : .WordStack
+               | #asByteStackInWidth(VALUE, 32)
+               })
+    )
+
+refund: _ => _
+
```

Comment:
As I said in the above (maybe it's my fault you didn't get it), the refund happens when you update an existing value to 0 in the storage (e.g., in this case, the caller's balance becomes 0). So, the functions that touch the storage, i.e., `approve`, `transfer`, and `transferFrom`, may result in a refund.

Reply:
yes, i got it in the end.  It would be nice to include the refund, too, in the spec, because it would show everybody that we are very serious about the last detail.